### DISCUSSION TITLE
Same group and sub group

### DISCUSSION THREAD

**Discussion by LeXXik** [jrouwe/JoltPhysics.js#65] [Q&A] [answered]
Two bodies that have the same group id and the same sub group id do not collide for me. Is it correct?
For example, I create a single group, with only 1 sub group, then assign a couple of dynamic bodies to it - they don't collide. I tried to enable the sub group collision with itself, but it didn't work either:
```js
filter.EnableCollision(0, 0);
```
The bodies start colliding as soon as I assign each to its own sub group, though.

**Comment by jrouwe** [comment, marked as answer]
If sub groups are the same, they will never collide (EnableCollision will assert if you do, but unfortunately this is not visible in the WASM version). See 'Detailed Description' of the algorithm here: https://jrouwe.github.io/JoltPhysics/class_group_filter_table.html

**Comment by LeXXik** [comment]
Ah, how did I miss that... Thanks!

**Reply by LeXXik** [reply]
Curious, if the groups examples could be improved. At the moment, I think it is used in constraints, where the groups pairs are explicitly disabled. Probably the correct way would be just to add them all to the same group, as we often do not know the amount of bodies ahead of time.

**Reply by jrouwe** [reply]
If you want none of the bodies to collide then that's an option. Usually if you have a long chain, you want the head to be able to collide with the tail, you just don't want each link to collide with the next link, so then you need to explicitly assign each body to a sub group. Note that if the current GroupFilterTable class is not suitable for your purposes, I can expose the interface to JS so you can override it and have your own implementation that does whatever suits you best (in C++ it's a virtual interface).

**Reply by LeXXik** [reply]
Right, that makes sense. This way only the direct neighbors won't collide, while the snek would still coillide with itself.

Regarding the GroupFilterTable - I think it is fine. The current implementation answers the need. I don't really have a complicated filtering scenario yet, so we can leave it as is until a need arises.

**Comment by lo-th** [comment]
Hi, standard implementation in physics engine use bits collisionGroup and collisionMask
is it possible to create a  GroupFilterTable that use this methode ?

https://lo-th.github.io/phy/#collision

**Reply by jrouwe** [reply]
I think the best way would be to create a custom ObjectLayerPairFilter implementation, see [this](https://github.com/jrouwe/JoltPhysics/issues/744) post. I think you will also need a BroadPhaseLayerInterface and ObjectVsBroadPhaseLayerFilter to go with that but I haven't made it possible to override them in JS yet. If I get the time, I'll create an example.

**Reply by lo-th** [reply]
great here is the default group in ammo, i use this rule for all engine
-1   : ALL
1    : BODY
2    : STATIC
4    : KINEMATIC
8    : DEBRIS
16   : SENSORTRIGGER
32   : NOCOLLISION
64   : GROUP0
128  : GROUP1 ...
also a simple function to disable contraint collision is great ( give same SubGroupID work )

**Reply by LeXXik** [reply]
I don't think the list is correct:
https://pybullet.org/Bullet/BulletFull/structbtBroadphaseProxy.html

**Reply by jrouwe** [reply]
I exposed the functionality and created an [example](https://jrouwe.github.io/JoltPhysics.js/alternative_collision_filtering.html):

https://github.com/jrouwe/JoltPhysics.js/blob/1032c7db6b6327e707d059af3cca5a73dcf8d391/Examples/alternative_collision_filtering.html#L24-L52
